tantar/src/main/java/org/opencadc/tantar/BucketValidator.java

```diff
@@ -588,82 +591,139 @@ public void updateArtifact(final Artifact artifact, final StorageMetadata storag
 
             @Override
             public boolean hasNext() {
-                if (storageMetadataIterator.hasNext()) {
+                if (artifactIterator.hasNext()) {
                     return true;
                 } else if (bucketPrefixIterator.hasNext()) {
-                    try {
-                        storageMetadataIterator =
-                                Subject.doAs(runUser, (PrivilegedExceptionAction<Iterator<StorageMetadata>>) () ->
-                                                                  storageAdapter.iterator(bucketPrefixIterator.next()));
-                    } catch (PrivilegedActionException exception) {
-                        throw new IllegalStateException(exception);
-                    }
+                    artifactIterator = iteratorDAO.storedIterator(bucketPrefixIterator.next());
                     return hasNext();
                 } else {
                     return false;
                 }
             }
 
             @Override
-            public StorageMetadata next() {
-                return storageMetadataIterator.next();
+            public Artifact next() {
+                return artifactIterator.next();
             }
         };
     }
 
     /**
-     * Iterate over the Artifact instances from the Storage Inventory database.  The consumer of this Iterator will
-     * assume that the StorageLocation is never null in the Artifact.
-     * <p/>
-     * Tests can override this for convenience.
+     * Check the StorageMetadata for a matching ObsoleteStorageLocation and if found
+     * delete the file from storage and delete the ObsoleteStorageLocation.
+     */
+    boolean isObsoleteStorageLocation(StorageMetadata storageMetadata) {
+        ObsoleteStorageLocation obsoleteStorageLocation;
+        try {
+            obsoleteStorageLocation = this.obsoleteStorageLocationDAO.get(storageMetadata.getStorageLocation());
+        } catch (Exception e) {
+            throw new IllegalStateException(
+                String.format("query failed for ObsoleteStorageLocation %s for file %s",
+                              storageMetadata.getStorageLocation().getStorageID().toASCIIString(),
+                              storageMetadata.artifactURI.toASCIIString()), e);
+        }
+        if (obsoleteStorageLocation != null && canTakeAction()) {
+            try {
+                delete(storageMetadata);
+                this.obsoleteStorageLocationDAO.delete(obsoleteStorageLocation.getID());
+                LOGGER.info(String.format("deleted obsolete file %s at %s",
+                                          storageMetadata.artifactURI.toASCIIString(),
+                                          obsoleteStorageLocation.getLocation().getStorageID().toASCIIString()));
+            } catch (Exception e) {
+                LOGGER.error(String.format("delete failed for obsolete file %s at %s",
```

Comment:
Aside from the same issue about correctness as above, this log message assumes that the `delete(storageMetadata)` was the action that failed, which might not be true. 
```
failed to cleanup obsolete {storageLocation}
```
More importantly, look at the now-emphasized text in my previous review comment concerning error handling.

Reply:
Updated to throw an exception if the cleanup fails.